src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp

```diff
@@ -353,7 +357,7 @@ void ShenandoahBarrierSetAssembler::load_at(MacroAssembler* masm, DecoratorSet d
     BarrierSetAssembler::load_at(masm, decorators, type, dst, src, tmp1, tmp_thread);
 
     if (ShenandoahBarrierSet::use_load_reference_barrier_native(decorators, type)) {
-      load_reference_barrier_native(masm, dst, src);
+      load_reference_barrier_native(masm, dst, src, (decorators & IN_NATIVE) == 0);
```

Comment:
Yeah, this is confusing. Zhengyu also stumbled over this. Notice that (decorators & IN_NATIVE) == 0 tests for 'is *not* native'. The point is that native-access is *always* uncompressed-oops, while accessing a referent is narrowOop or oop depending on UseCompressedOops. Hence the distinction. If you have a good suggestion on how to make this less confusing, I'd appreciate it.

Reply:
But that's the thing that gets my head spinning. Why do we call into `lrb_native` for referents? This contradicts the idea that "native-access is always uncompressed-oops". I think this tries to overload "native" with more meaning that it is equipped to carry. 

I think at very least it should say:

```
    if (ShenandoahBarrierSet::use_load_reference_barrier_native(decorators, type)) {
      // API impedance: when used on native refs, lrb-native necessarily works with full oops,
      // but when used for weak/phantom refs, it might need to work with narrow oops.
      // Therefore, we need to ask barrier code to look back at UseCompressedOops and
      // decide, when lrb-native is not IN_NATIVE. TODO: Resolve this API impedance.
      bool maybe_narrow_oop = (decorators & IN_NATIVE) == 0;
      load_reference_barrier_native(masm, dst, src, maybe_narrow_oop);
    } else {
      load_reference_barrier(masm, dst, src);
    }
```

Reply:
I thought about it today. This whole idea of 'LRB-native' is flawed. What it does is prevent resurrection of objects when loading from a field or off-heap-location that is 'weak' or 'phantom'. This has nothing to do with -native. It has to do with the field being not-strong. For this reason I think we should call that variant of LRB something like LRB-weak instead. This warrants a larger reshuffling that I'd do either before or after this change goes in. I'd probably also merge our 3(!) different runtime LRB impls into one, with templated path to prevent resurrection, etc. I guess the 2 interpreter LRB entries can also be unified, and differ only in the target entry being called.

The only distinction for which IN_NATIVE is relevant is for figuring out whether or not the reference is compressed or not. That is all.

Reply:
OK, so that would resolve the `TODO` I suggested in the code comment above, right? If so, we can just put in the code comment for now.